### Clase 3

In [1]:
import csv
import os
import numpy as np
import pandas as pd

Cargamos el conjunto de datos como una listas de listas.

In [2]:
path = os.path.join('..','Datasets','GLT_filtrado.csv')

conjunto = [[],[],[],[],[],[],[]]

with open(path, 'r') as f:
    data = list(csv.reader(f))

    labels = data[0]

    for i in data[1:]:
        conjunto[0].append(i[0] if i[0]!='' else np.nan)
        conjunto[1].append(float(i[1]) if i[1]!='' else np.nan)
        conjunto[2].append(float(i[2]) if i[2]!='' else np.nan)
        conjunto[3].append(i[3] if i[3]!='' else np.nan)
        conjunto[4].append(i[4] if i[4]!='' else np.nan)
        conjunto[5].append(i[5] if i[5]!='' else np.nan)
        conjunto[6].append(i[6] if i[6]!='' else np.nan)

Contestemos las siguientes preguntas:

* ¿Tipo de dato de los atributos?
* ¿Cuántas entradas hay?
* ¿Cuántas ciudades distintas?

In [3]:
# Total de datos
total = len(data[1:])
print(f'Total de datos: {total}')

# El índice 3 de `conjunto`` corresponde a las ciudades
ciudades_unicas = np.unique(conjunto[3])
print(f"Tenemos {len(ciudades_unicas)} ciudades únicas.")



Total de datos: 219575
Tenemos 100 ciudades únicas.


¿Cuantos valores faltantes hay en el atributo AverageTemperature?

In [4]:
# El índice 1 de `conjunto` corresponde a AverageTemperature
faltan = np.count_nonzero(np.isnan(conjunto[1]))
print(f"Hay {faltan} valores faltantes en el atributo 'AverageTemperature'.")
print(f"Los datos faltantes representan el {faltan*100/total:.2f}%")
print('-'*80)

# Verificamos valores faltantes para los demás atributos
for i in range(7):
    if i==1:
        continue
    print(f"Hay {np.count_nonzero(pd.isna(conjunto[i]))} valores faltantes en el atributo '{labels[i]}'")

Hay 27581 valores faltantes en el atributo 'AverageTemperature'.
Los datos faltantes representan el 12.56%
--------------------------------------------------------------------------------
Hay 0 valores faltantes en el atributo 'Date'
Hay 27581 valores faltantes en el atributo 'AverageTemperatureUncertainty'
Hay 0 valores faltantes en el atributo 'City'
Hay 0 valores faltantes en el atributo 'Country'
Hay 0 valores faltantes en el atributo 'Latitude'
Hay 0 valores faltantes en el atributo 'Longitude'


Como vemos, los valores faltantes para AverageTemperature (así como para AverageTemperatureUncertainty) representan el 12.56%

En primera instancia no sabemos la ubicación de los valores faltantes en el conjunto de datos, es decir, hasta este punto no sabemos a qué fechas y/o ciudades les están faltando valores. El porcentaje que representan los valores faltantes es significativo pues afecta la distribución de la temperatura generando una importante dispersión en los datos. Se espera que esta distribución sea continua y suave en el tiempo pues esta es la naturaleza física del cambio en temperatura, es decir, sin saltos abruptos. Por ello, necesitamos aplicar una estrategia que nos permita rellenar de forma coherente los huecos faltantes para reducir el sesgo y dispersión, la cual se verá a continuación.

In [5]:
def llenar_faltantes(ar):
    ...

In [22]:
import numpy as np

p = [ np.nan, 5, 6, np.nan ]

p = np.array(p)  # Aseguramos que es un array de NumPy
d = np.copy(p)   # Copia para almacenar los valores corregidos

nan_indices = np.where(np.isnan(d))[0]  # Encuentra los índices donde hay NaN

for i in nan_indices:
    anterior = d[i-1] if i > 0 else 0  # Si es el primer elemento, usa 0

    # Buscar el siguiente valor no NaN
    j = i
    while j < len(d) and np.isnan(d[j]):
        j += 1
    siguiente = d[j] if j < len(d) else 0  # Si llega al final, usa 0

    d[i] = (anterior + siguiente) / 2  # Promedio

print(d)


[2.5 5.  6.  3. ]


In [21]:
# original_1
p = [ np.nan, 5, 6, np.nan ]

# original_2
# p = [ np.nan, np.nan, 24, 28.9 ]

# p = [np.nan, np.nan,24, np.nan, np.nan]
# p = [12,np.nan, np.nan, 15,np.nan, 16]

d = p
for i in range(len(p)):
    if p[i] is np.nan:
        if i == 0:
            anterior = 0
        else: 
            anterior = p[i-1]

        j = i
        while p[j] is np.nan:
            if j == len(p)-1:
                siguiente = 0
                break
            else:
                siguiente = p[j+1]
                j+=1

        prom = (anterior + siguiente) / 2
        d[i] = prom
        
d

[2.5, 5, 6, 3.0]

In [7]:
p = [np.nan, np.nan,np.nan,24, np.nan,np.nan,np.nan]
p = [np.nan, np.nan, np.nan,np.nan,100, np.nan, np.nan,np.nan]

d = p
flag = False
i = 0

while True:
    print(i)
    if p[i] is np.nan:
        if i==0:
            anterior = 0
        else: 
            anterior = p[i-1]

        j = i
        while p[j] is np.nan:
            if j == len(p)-1:
                siguiente = 0
                flag = True
                p.reverse()
                k=i
                i=0
                break
            else:
                siguiente = p[j+1]
                j+=1

        prom = (anterior + siguiente) / 2
        d[i] = prom

    if i == len(p) - 1 and flag == False:
        break
    if i == len(p) - k - 1 and flag == True:  # i==7-4-1 = 2
        p.reverse()
        break

    i+=1
        
d

0


NameError: name 'k' is not defined